### PR TITLE
#167765333 admin can add a new user

### PR DESCRIPTION
**Pull request by victor-nach**
## What does this PR do?
- Allows an admin to create a new user
## Description of Task to be completed?
- Write controller function in (src/controllers/userController.js) that would use the user models to create a new user
- Create standard response messages in (src/utils/response.js) to be used by user controller
- Write unit tests
- Document using swagger

## How should this be manually tested?
- Run ``npm run test`` to test the branch
- send a post request containing the relevant user fields to POST /api/v1/users

```
- Sample request body
{
	"firstName":"victor",
	"lastName": "nach",
	"userName": "lazydev",
	"email": "myemail@gmail.com",
	"password": "password",
	"phoneNumber": "09034343434"
}
```

## What are the relevant pivotal tracker stories?
- #167765333

## Implementation plan
- https://docs.google.com/document/d/1uZsbldxw0rMNMtuKPRlWWWW71vqsq9dXRUSFG2jsuJg/edit#heading=h.oucxq2oo74zs

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,7 +5,7 @@
   "main": "index.js",
   "scripts": {
     "start": "node ./src/bin/www",
-    "devstart": "babel-watch ./src/index.js",
+    "devstart": "babel-watch --exec babel-node ./src/index.js",
```
Comment (debelistic) [inline package.json]: is `babel-watch` not handing things properly here. It should be enough to without using `babel-node`

File: src/routes/api/index.js [hunk 2]
```diff
@@ -1,18 +1,6 @@
...
-            errors: Object.keys(err.errors).reduce(function(errors, key) {
-                errors[key] = err.errors[key].message;
-                return errors;
-            }, {})
-        });
-    }
-
-    return next(err);
-});
+router.use('/', userRoutes);
```
Comment (debelistic) [inline src/routes/api/index.js]: Since es6 is available, why don't you switch to import/export instead. I'm referring to lines 1 and 6

File: package.json [hunk 3]
```diff
@@ -5,14 +5,17 @@
...
     "build": "rm -rf dist && mkdir dist && babel src -s -d dist",
-    "test": " set NODE_ENV=test&& mocha -r @babel/register --recursive src/tests/ --exit",
+    "test": " set NODE_ENV=test && npm run reset:db && mocha -r @babel/register --recursive src/tests/ --exit --timeout 10000",
     "generate-lcov": "nyc report --reporter=text-lcov > lcov.info",
```
Comment (uggoprince) [inline package.json:11]: why is this here?
coveralls alone is okay

**Review by uggoprince — COMMENTED** [review]
export controller like this `export default new UserController();`, therefore there'll be no need for `export` statement at the top of the class and you wont need to create and instance (object) of it in the` route` and no need for `static methods`

Use error code `400` in place of `409`

You don't need to use `return` statements in your controller. `res` object is enough, and you do it like this `res.status(200).send({...});` not `res.status(200).json({...});`

`config.js` for database needs to be in a `config folder` inside database folder.

database `field` should be in `snake case` (underscore).

You need to have a separate `json file` for User swagger documentation. Having one file for swagger
documentation will make it hard to maintain. And `public/api-docs` should not be in the `src` folder, check `Simulations Project` pdf that was sent and see. @voke1

At the end your controller looks like this:
```
class UserController {
/**
 * @name CreateUser
 * @description Allows an admin add a new user
 * @param {object} req The request object
 * @param {object} res The response object
 * @returns {object} The API response
 */
  async createUser(req, res) {
    try {
      const userData = { ...req.body };
      const { dataValues } = await models.User.create(userData);
      delete dataValues.password; // remove sensitive data from returned object
      res.status(201).send({
        status: 201,
        message: 'User successfuly created',
        data: { ...dataValues }
      });
    } catch (error) {
      if (error.name === 'SequelizeUniqueConstraintError') {
        res.status(400).send({
          status: 400,
          error: 'Unsuccesful, user already exists, kindly use a different email and/or username.',
        });
      }
      res.status(500).send({ error: error.message });
    }
  }
}

export default new UserController();
```

File: package.json (hunk 3 above)
Comment (victor-nach) [reply package.json:11]: I don't understand what you're asking, I didn't touch the coveralls, I only added instructions to clean the database before running the tests.

**Review by debelistic — CHANGES_REQUESTED** [review]
*index.js*
- you declared import route twice

*routes/index.js*

- you declared api twice
- line 3 and line 8 should use es6

*reponse in controller*

- destructure username, email, password, firstname, lastname and phonenumber from the data no need to add the remaining
- use spell check extension to keep track of spellings

*controllers/index.js*

- contains another class of usercontroller
- should only be importing and exporting controllers e.g
`export { default as userController } from './users';`. You can then import it anywhere you need it as 
`import { userController } from '../../controllers';`

PS: eslint should flag it initially cause that's the only export but once we add more

*package.json*
- our devstart script doesn't need babel-node to execute, its redundant. You can see babel-watch doc, if you're not sure how it works
- you preceded both migrate and reset:db with npm run, there's no need.
- you installed the nodemon, you can uninstall it. babel-watch, takes care of what ever we need it for. Those scripts are failing at the moment. For reset:db use hyphen instead

*test*
test script should be
`    "test": "set NODE_ENV=test&& npm run reset-db && mocha -r @babel/register --recursive src/tests/ --exit",`. Leave a space after test would result to `test ` in the console, I don't think tests would be running at the moment.
- specify test db in env file

*bycrpt*
use bcrypt to encrypt password before saving to db
cc: https://stackoverflow.com/questions/34120548/using-bcrypt-with-sequelize-model

**Comment by victor-nach** [comment]
@debelistic 
-  Any specific reason you want us to use 400 instead of 409  'user already exists' error? 400 may not be an appropriate error in that case.
-  I used static methods in the classes, so those don't need to create an instance of an object to make use of them
- Also, any specific reason you want us to use snake case for the database, sequelize already helps you maintain case sensitivity in your SQL database, or is there a project guide for that because I didn't see any.

File: src/routes/api/index.js (hunk 2 above)
Comment (victor-nach) [reply src/routes/api/index.js]: most of the es5 code there was what was initially there, I'll remove this one too thanks.

**Comment by victor-nach** [comment]
- I would the bcrypt
- nodemon has already been removed
- the test scripts work, have you tested them yourself?

File: src/tests/user.spec.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
+import chai from 'chai';
```
Comment (voke1) [inline src/tests/user.spec.js:1]: Good work @victor-nach I advise you change `user.spec.js` to `user.js` as a naming convention

File: package.json (hunk 3 above)
Comment (uggoprince) [reply package.json:11]: i mean `generate-lcov`

Comment (uggoprince) [reply package.json:11]: I also forgot this: You need to create two databases, one for `development` and one for `test`. During tests we set database_url to that of `test` and at the end of all tests drop all its tables.

**Comment by uggoprince** [comment]
> @debelistic
> 
> * Any specific reason you want us to use 400 instead of 409  'user already exists' error? 400 may not be an appropriate error in that case.
> * I used static methods in the classes, so those don't need to create an instance of an object to make use of them
> * Also, any specific reason you want us to use snake case for the database, sequelize already helps you maintain case sensitivity in your SQL database, or is there a project guide for that because I didn't see any.

Sequelize does what you tell it to do. And that's before migrations are created. So you can decide what case to use before creating models and migrations. when you run migrations it creates as you specified before. And is't good convention using snake case for `field names`

`400` or `422` is widely used for duplicate entry not `409`. `409` is hardly used  and read about conflicts mostly for `PUT` and uploading of files.

When you export an instance e.g `export default new ClassName();` then there's no need creating an instance where you will import it. And there will be no need for `static methods`.

**Comment by victor-nach** [comment]
- javascript uses camelcase conventionally, if we decide to use snake case even es-lint would flag that, I don't also see any advantage of using camel case.
- for the error codes, I think we should involve other members of the team.
- using a static method goes the same way too, also no need to create an instance when importing and also to create an instance when exporting. I also feel the syntax is cleaner that way, both methods are valid and would work, choosing any of them is opinionated, I think we should involve other members of the team as well.

File: src/index.js [hunk 5]
```diff
@@ -46,6 +47,7 @@ if (!isProduction) {
 }
 
 
+app.use('/api/v1', router);
```
Comment (voke1) [inline src/index.js]: I suggest you use `app.use('/', router);` instead of `app.use('/api/v1', router);` and in the `src/routes/api/index.js` file you use `router.use('/api/v1', userRoutes);` This way, we are able to connect all routes in the entry file and filter routes in the `src/routes/api/index.js` file.

**Review by voke1 — CHANGES_REQUESTED** [review]

**Comment by uggoprince** [comment]
> * javascript uses camelcase conventionally, if we decide to use snake case even es-lint would flag that, I don't also see any advantage of using camel case.
> * for the error codes, I think we should involve other members of the team.
> * using a static method goes the same way too, also no need to create an instance when importing and also to create an instance when exporting. I also feel the syntax is cleaner that way, both methods are valid and would work, choosing any of them is opinionated, I think we should involve other members of the team as well.

Yes eslint will flag and all we have to do is disable that rule.

For the static method vs instance you getvsame result but the code is cleaner with instance procedure. Most Andela fellows I've met make use of the instanciation pattern.

File: src/index.js (hunk 5 above)
Comment (victor-nach) [reply src/index.js]: ok, would update that.

File: src/tests/user.spec.js (hunk 4 above)
Comment (victor-nach) [reply src/tests/user.spec.js:1]: ok, would do that, thanks

File: package.json (hunk 3 above)
Comment (victor-nach) [reply package.json:11]: - ok I see, but I'm not the one that added the geerated l-cov.
- For the test databse url there is already provision for that in the sequelize config file, it just needs a little adjustment, that may be out of the scope of this branch.

**Review by misanj — APPROVED** [review]
Great job @victor-nach looks good to me....just implement some of the suggestions from other team mates.

**Comment by victor-nach** [comment]
@debelistic  
- concerning destructuring the required fields from the req.body, it really doesn't matter what fields you pass in, sequelize would only work with the fields specified in the user model. I wrote it this way so that if in the future we decide to change the fields required, we only have to edit it from the user model, and also because it's shorter. So what do you think about this?

**Comment by voke1** [comment]
@victor-nach Are you able to run `npm run test` successfully? I'm encountering some errors. see below:
```
C:\Users\USER\Documents\latestFolders\phoenix-bn-backend\node_modules\.bin\sequelize:2
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
          ^^^^^^^

SyntaxError: missing ) after argument list
    at new Script (vm.js:80:7)
    at createScript (vm.js:274:10)
    at Object.runInThisContext (vm.js:326:10)
    at Module._compile (internal/modules/cjs/loader.js:664:28)
    at Module._extensions..js (internal/modules/cjs/loader.js:712:10)
    at Object.newLoader [as .js] (C:\Users\USER\Documents\latestFolders\phoenix-bn-backend\node_modules\pirates\lib\index.js:104:7)
    at Module.load (internal/modules/cjs/loader.js:600:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:539:12)
    at Function.Module._load (internal/modules/cjs/loader.js:531:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:754:12)```

**Comment by debelistic** [comment]
@victor-nach the destructing, I was talking about is on the `response` sent back on. Also, the user actually needs the password.

File: src/controllers/users.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+export default class UserController {
+/**
+ * @name CreateUser
+ * @description Allows an admin add a new user
+ * @param {object} req The request object
+ * @param {object} res The response object
+ * @returns {object} The API response
+ */
+  static async createUser(req, res) {
+    try {
```
Comment (anneKay) [inline src/controllers/users.js]: I would like for you to check if the user exists first before proceeding with creating the user, so if user exists, just return the response with a status of 409, and if not then we can go ahead to create the user.

File: package.json [hunk 7]
```diff
@@ -5,14 +5,17 @@
...
     "coveralls": "nyc report --reporter=text-lcov | coveralls",
     "codeclimate": "codeclimate-test-reporter > lcov.info",
     "coverage": "nyc npm test && npm run generate-lcov && npm run coveralls && npm run codeclimate",
-    "generate-report": "nyc --reporter=lcov --reporter=text-lcov npm test"
+    "generate-report": "nyc --reporter=lcov --reporter=text-lcov npm test",
+    "sequelize": "./node_modules/.bin/babel-node ./node_modules/.bin/sequelize $*",
+    "migrate": "npm run sequelize db:migrate",
```
Comment (voke1) [inline package.json]: For the migrate script, just `'sequelize db:migrate'` should do

File: package.json [hunk 8]
```diff
@@ -5,14 +5,17 @@
...
+    "reset:db": "npm run sequelize db:migrate:undo:all && npm run migrate"
```
Comment (voke1) [inline package.json]: Also for reset:db script `sequelize db:migrate:undo:all && npm run migrate"` should do

File: package.json (hunk 7 above)
Comment (victor-nach) [reply package.json]: that won't work unless sequelize is installed globally

**Review by misanj — APPROVED** [review]
Well done guys! Looks good to me

File: src/middlewares/userMiddlewares.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ * @param {object} req The request object
+ * @param {object} res The response object
+ * @param {object} next The response object
+ * @returns {object} The API response or next()
+ */
+  static async checkUserExists(req, res, next) {
+    try {
+      const { email } = req.body;
+      const data = await models.User.findOne({ where: { email, } });
+      console.log(data);
```
Comment (anneKay) [inline src/middlewares/userMiddlewares.js]: Remove the console.log

File: src/middlewares/userMiddlewares.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+      if (!data) return next();
+      return res.status(409).json({
+        status: 409,
+        error: 'Unsuccesful, user already exists, kindly use a different email and/or username.',
```
Comment (anneKay) [inline src/middlewares/userMiddlewares.js]: Are we making userName and email unique?

File: src/migrations/20190827125917-create-user.js [hunk 11]
```diff
@@ -30,24 +44,34 @@ module.exports = {
...
     phoneNumber: {
       allowNull: false,
       type: Sequelize.STRING,
       validate: {
-        not: ['[a-z]', 'i']
-      }
+        notNull: true,
+      },
+    },
+    isAdmin: {
```
Comment (anneKay) [inline src/migrations/20190827125917-create-user.js:58]: Please don't validate this field to be `not null` instead give it a default value of false. If the user wants to add an admin, he can specify that and if not, he doesn't have to

File: src/tests/user.spec.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+      .post(`${endPoint}/auth/signup`)
+      .send(user)
+      .end((_err, res) => {
+        expect(res).to.have.status(201);
+        expect(res.body.status).to.be.equal(201);
+        expect(res.body).to.have.property('data');
+        expect(res.body.data).to.be.a('object');
+        expect(res.body.data).to.have.property('id');
+        expect(res.body.data.id).to.be.a('number');
+        expect(res.body.data.id % 1).to.be.equal(0);
```
Comment (anneKay) [inline src/tests/user.spec.js:31]: What is this testing for?

File: src/middlewares/userMiddlewares.js (hunk 10 above)
Comment (victor-nach) [reply src/middlewares/userMiddlewares.js]: what do you suggest?, only email should be unique?

File: src/migrations/20190827125917-create-user.js (hunk 11 above)
Comment (victor-nach) [reply src/migrations/20190827125917-create-user.js:58]: ok

File: src/tests/user.spec.js (hunk 12 above)
Comment (victor-nach) [reply src/tests/user.spec.js:31]: The id is the primary key which we need to be returned. I'm testing that the response body has a field 'id' and that it is a number and also an integer. The tests pass as is.

File: src/controllers/UserController.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+ * @param {object} req The request object
+ * @param {object} res The response object
+ * @returns {object} The API response
+ */
+  static async createUser(req, res) {
+    try {
+      const userData = { ...req.body };
+      // hash the password before storing in the db
+      userData.password = Utils.hashPassword(userData.password);
+      const { dataValues } = await models.User.create(userData);
```
Comment (V-84) [inline src/controllers/UserController.js]: create a userService class that abstracts this.

File: src/middlewares/userMiddlewares.js [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+ * @description Checks if a user exists in the database
+ * @param {object} req The request object
+ * @param {object} res The response object
+ * @param {object} next The response object
+ * @returns {object} The API response or next()
+ */
+  static async checkUserExists(req, res, next) {
+    try {
+      const { email } = req.body;
+      const data = await models.User.findOne({ where: { email, } });
```
Comment (V-84) [inline src/middlewares/userMiddlewares.js]: same abstraction implies here

**Review by voke1 — APPROVED** [review]

File: src/controllers/UserController.js [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+ * @name CreateUser
+ * @description Allows an admin add a new user
+ * @param {object} req The request object
+ * @param {object} res The response object
+ * @returns {object} The API response
+ */
+  static async createUser(req, res) {
+    try {
+      const userData = { ...req.body };
+      // hash the password before storing in the db
```
Comment (V-84) [inline src/controllers/UserController.js]: remove inline comments

File: src/controllers/UserController.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+      userData.password = Utils.hashPassword(userData.password);
+      const { dataValues } = await models.User.create(userData);
+      delete dataValues.password; // remove sensitive data from returned object
```
Comment (V-84) [inline src/controllers/UserController.js]: remove inline comments

**Review by V-84 — CHANGES_REQUESTED** [review]
Nice work victor.
However, your code could be slightly improved.
see more comment below.
Reach out to me to explain the abstraction if you dont understand.
Also, please write unit tests for your utils class

**Review by debelistic — APPROVED** [review]

**Review by V-84 — APPROVED** [review]

File: src/middlewares/userMiddlewares.js [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+ * @returns {object} The API response or next()
+ */
+  static async checkUserExists(req, res, next) {
+    try {
+      const { email } = req.body;
+      const data = await userServices.getUserByEmail(email);
+      if (!data) return next();
+      return res.status(409).json({
+        status: 409,
+        error: 'Unsuccesful, user already exists, kindly use a different email and/or username.',
```
Comment (anneKay) [inline src/middlewares/userMiddlewares.js]: If we are no longer making the username unique, then update this error message accordingly
